File: server/src/collectors/pm25Japan.js
/**
 * PM2.5 Japan collector (pm25-japan).
 *
 * The MOE Soramame / AEROS air-quality network is the authoritative source and
 * is already mirrored by the dedicated soramame/AEROS collectors. The public
 * soramame.env.go.jp site is a JavaScript SPA — the historical key-free CSV
 * paths (data/sokutei/NoudoFile/...) now return the SPA HTML shell, not data,
 * so there is no usable key-free PM2.5 JSON/CSV endpoint for this source ID.
 *
 * Per the no-fabrication rule this returns an honest EMPTY FeatureCollection
 * (PM2.5 belongs to the existing soramame source). We still do a best-effort
 * reachability probe so the source surfaces an honest live/unavailable state.
 *
 * NOTE: a key-free Soramame PM2.5 JSON endpoint is UNVERIFIED / believed
 * absent; AEROS API access requires registration (no public key here).
 */

import { fetchHead } from './_liveHelpers.js';

const PROBE = 'https://soramame.env.go.jp/';

export default async function collectPm25Japan() {
  const reachable = await fetchHead(PROBE, { timeoutMs: 8000 }).catch(() => false);
  return {
    type: 'FeatureCollection',
    features: [],
    _meta: {
      source: reachable ? 'pm25_japan_no_public_endpoint' : 'pm25_japan_unavailable',
      fetchedAt: new Date().toISOString(),
      recordCount: 0,
      live: false,
      description: 'PM2.5 Japan — no key-free machine-readable endpoint (covered by soramame/AEROS); honest empty',
    },
  };
}

File: native/collectors/sources/pm25_japan.c
/* collectors/environment/sources/pm25_japan.c
 * FEED source — port of server/src/collectors/pm25Japan.js.
 * PM2.5 has no key-free machine-readable endpoint (covered by soramame/AEROS).
 * Node returns an honest EMPTY FeatureCollection (zero features) after a
 * best-effort reachability probe. Faithful port: probe, emit NOTHING.
 * RULE 8: never fabricate — honest empty is correct here. */
#include "source.h"
#include "lib/probe.h"
#include <stdio.h>

#define PROBE "https://soramame.env.go.jp/"

static int run(const source_ctx *ctx, intel_sink *sink) {
  (void)sink;
  int reachable = probe_head(ctx->http, PROBE);
  /* Node emits zero features regardless (no public endpoint) — emit nothing. */
  fprintf(stderr, "[pm25-japan] %s; no public endpoint, emitted 0\n",
          reachable ? "pm25_japan_no_public_endpoint" : "pm25_japan_unavailable");
  return 0;
}

static const source_def pm25_japan_def = {
  .id = "pm25-japan", .collector = "environment",
  .name = "PM2.5 Distribution", .name_ja = "PM2.5分布予測",
  .update_interval_sec = 3600, .run = run,
};
REGISTER_SOURCE(pm25_japan_def)
